drivers/serial/uart_stm32.c

```diff
@@ -1795,11 +1824,27 @@ static int uart_stm32_async_rx_enable(const struct device *dev,
 	/* Enable RX DMA requests */
 	uart_stm32_dma_rx_enable(dev);
 
-	/* Enable IRQ IDLE to define the end of a
-	 * RX DMA transaction.
-	 */
-	LL_USART_ClearFlag_IDLE(usart);
-	LL_USART_EnableIT_IDLE(usart);
+#if HAS_RTO
+	if (IS_UART_RTO_INSTANCE(usart)) {
+		/* Use hardware RTO interrupt */
+		uint32_t timeout_in_bits = timeout * data->uart_cfg->baudrate / 1000000;
```

Comment:
`timeout == -1` stands for no timeout (`SYS_FOREVER_US`). It such a case RTO interrupt should simply not be armed. For simplicity, maybe treat any negative timeout value as an no timeout configuration request.

Another issue here actually. Timeout in is microsecond.  If use uses a timeout of 1 or few second, this 32bit multiplication may overflow before the division. I would suggest to use `uint64_t` for the calculus.